bookshelf/routes/web.php

```diff
@@ -0,0 +1,199 @@
+        'books' => $query->documents(),
+        'pageSize' => $pageSize,
+    ]);
+});
+
+
+// [START add]
+$router->get('/books/add', function () {
+    return view('form', [
+        'action' => 'Add',
+        'book' => [],
+    ]);
+});
+
+$router->post('/books/add', function (Request $request) use ($collection, $gcsBucket) {
+    $files = $request->files;
+    $bookData = $request->request->all();
+    $image = $files->get('image');
+    if ($image && $image->isValid()) {
+        $f = fopen($image->getRealPath(), 'r');
+        $object = $gcsBucket->upload($f, [
+            'metadata' => ['contentType' => $image->getMimeType()],
+            'predefinedAcl' => 'publicRead',
+        ]);
+        $bookData['image_url'] = $object->info()['mediaLink'];
+    }
+
+    // Create the book
+    $bookRef = $collection->newDocument();
+    $bookRef->set($bookData);
+
+    return redirect("/books/" . $bookRef->id());
+});
+// [END add]
+
+// [START show]
+$router->get('/books/{bookId}', function ($bookId) use ($collection) {
+    # [START firestore_client_get_book]
+    $bookRef = $collection->document($bookId);
+    $snapshot = $bookRef->snapshot();
+    # [END firestore_client_get_book]
+
+    if (!$snapshot->exists()) {
+        return new Response('', Response::HTTP_NOT_FOUND);
+    }
+
+    return view('show', ['book' => $snapshot]);
+});
+// [END show]
+
+// [START edit]
+$router->get('/books/{bookId}/edit', function ($bookId) use ($collection) {
+    $bookRef = $collection->document($bookId);
+    $snapshot = $bookRef->snapshot();
+
+    if (!$snapshot->exists()) {
+        return new Response('', Response::HTTP_NOT_FOUND);
+    }
+
+    return view('form', [
+        'action' => 'Edit',
+        'book' => $snapshot,
+    ]);
+});
+
+$router->post('/books/{bookId}/edit', function (Request $request, $bookId) use ($collection, $gcsBucket) {
+    $bookRef = $collection->document($bookId);
+    $snapshot = $bookRef->snapshot();
+
+    if (!$snapshot->exists()) {
+        return new Response('', Response::HTTP_NOT_FOUND);
+    }
+
+    // Get book data from the request object
+    $bookData = $request->request->all();
+    $bookData['id'] = $bookId;
+
+    // [START add_image]
+    $files = $request->files;
+    $image = $files->get('image');
+    if ($image && $image->isValid()) {
+        $f = fopen($image->getRealPath(), 'r');
+        $object = $gcsBucket->upload($f, [
+            'metadata' => ['contentType' => $image->getMimeType()],
+            'predefinedAcl' => 'publicRead',
+        ]);
+        $bookData['image_url'] = $object->info()['mediaLink'];
+    }
+    // [END add_image]
+
+    // Create array to update the Firestore document
+    $updateData = [];
+    foreach ($bookData as $key => $value) {
+        $updateData[] = ['path' => $key, 'value' => $value];
+    }
+
+    if ($bookRef->update($updateData)) {
+        return redirect("/books/$bookId");
+    }
+
+    return new Response('Could not update book');
```

Comment:
- Will this will ever be executed? It looks like either a result will be returned from `update` or an exception would be bubbled up from the transport. Perhaps we should instead be catching exceptions from `update`/ `upload` and then return this response?
- What is the flow like on the user's end if this response is returned?
- Should it return an error code?
- Why don't we return responses like this for the other actions if something fails?

Reply:
I wasn't sure if there was a case where `update` would just return `false` with no Exception. For this app, it's sufficient for the exception to be thrown. Other points are solid. Removed.